Generate shell completion scripts

Bouke opened the pull request:
This is in addition to the work done in [PR#703][1]. The status of this PR is WIP and pushed to get feedback from the maintainers.

I've made a few changes to the public API of ArgumentParser to get all the meta data required to generate the scripts, as well as the descriptions that zsh shows. For the SwiftPM tools, this means no duplication of the arguments anymore inside SwiftPM.

Things left to tackle:
* [x] Support positional arguments
* [x] Cleanup the code along with some documentation

Not handled in this PR, requires further discussion:
* Compiler flags are still duplicated; should we add some sort of "discovery" based on the output of `swift -h`? -- What to do with flags we cannot handle (nicely), like `-l<name>`?
* How to include this in the toolchain?
* How to generate the full completion script, including setup code? (see `_swift` functions in the current scripts under `Utilities/{bash,zsh}`.

[1]: https://github.com/apple/swift-package-manager/pull/703

aciidb0mb3r commented on Sources/Utility/ArgumentParser.swift:
```diff
@@ -193,15 +214,15 @@ public final class PositionalArgument<Kind>: ArgumentProtocol {
 ///
 /// Note: Only used for argument parsing purpose.
-fileprivate final class AnyArgument: ArgumentProtocol, CustomStringConvertible {
+public final class AnyArgument: ArgumentProtocol, CustomStringConvertible {
```
Why do you need AnyArgument to be public? See the note, it was only meant to be used by argument parser internally

Bouke replied:
In order to generate those scripts, meta information is required about the possible arguments, usage, overview and sub parsers. What do you propose where this information should come from?

aciidb0mb3r replied:
The script generators should be a part of ArgumentParser so this doesn't need to be public

aciidb0mb3r commented on Sources/Commands/SwiftPackageTool.swift:
```diff
@@ -392,12 +395,29 @@ public class SwiftPackageTool: SwiftTool<PackageToolOptions> {
     }
 }
 
+enum Shell: String, StringEnumArgument {
```
This should belong in `ArgumentParser` and not in swift tools because we want to be able to generate completion scripts for any parser instance.

aciidb0mb3r commented on Sources/Utility/ArgumentParser.swift:
```diff
@@ -29,6 +29,19 @@ public enum ArgumentParserError: Swift.Error {
+    case none
+    case unspecified
+    case filename
+    case values([(value: String, description: String)])
```
What is value and description?

Bouke replied:
Yeah those could definitely be improved. Take for example `swift package init` which can take a `--type` argument. The possible values for this argument along with a description would be written like this:

    let completion = ShellCompletion.values([
        ("empty", "generates an empty project"),
        ("executable", "generates a project for a cli executable"),
        ("library", "generates a project for a dynamic library"),
        ("system-module", "generates a project for a system library"),
    ])

aciidb0mb3r replied:
Ah I see

aciidb0mb3r commented on Sources/Commands/GenerateShellScript.swift:
```diff
@@ -0,0 +1,328 @@
+}
+
+
+// MARK:- BASH
```
All these methods should be in argument parser so completion can be generated for any parser. We should aim for a good API, maybe something which looks like this:
```swift
let parser = ArgumentParser(...)
let stream = ...
try parser.generateCompletion(for: .zsh, on: stream)
try parser.generateCompletion(for: .bash, on: stream)
```

Bouke replied:
That is a great point. However at the moment ArgumentParser isn't sufficient to represent the arguments of the `swift` binary; i.e. both the 3 tools and the compiler. So while I can agree with you it would be nice to generate scripts for any ArgumentParser, we need a solution for the "root" argument parser.

aciidb0mb3r commented on Utilities/bash/completions:
```diff
@@ -1,147 +1,564 @@
 _swift() 
 {
-    declare -a cur prev shared_options
+    declare -a cur prev
```
Lets not update these files with this feature

aciidb0mb3r replied:
(as in we can do it in subsequent PRs)

Bouke commented on Sources/Commands/GenerateShellScript.swift:
```diff
@@ -0,0 +1,328 @@
+    root.add(subparser: "test", parser: SwiftTestTool(args: []).parser)
+
+    // Swift compiler flags
+    _ = root.add(option: "-assert-config", kind: String.self, usage: "Specify the assert_configuration replacement. Possible values are Debug, Release, Unchecked, DisableReplacement.")
```
We currently [already do that][0]. So my idea is to either generate them (like in this PR), or include them as static code. I would rather not lose the ability to complete compiler arguments.

[0]: https://github.com/apple/swift-package-manager/blob/914931d932e63ddeeaa1ddb59f4b8ca42a125e5e/Utilities/zsh/_swift#L122-L161

aciidb0mb3r replied:
I agree that they're helpful but if we're going to do compiler options, we should have it in a separate PR

Bouke commented on Sources/Commands/Options.swift:
```diff
@@ -42,4 +42,25 @@ public class ToolOptions {
+/// Parser conformance for AbsolutePath.
+extension AbsolutePath: ArgumentKind {
+    public init?(arg: String) {
+        self.init(arg, relativeTo: currentWorkingDirectory)
```
What would be a better place for this? All this is saying that `AbsolutePath` conforms to the `ArgumentKind` protocol. Incorrect usage of any API should be discouraged. I see two options going forward; 1. add a note about the API or 2. introduce an additional type that conforms to ArgumentKind, maybe named `PathArgument`.

aciidb0mb3r replied:
A PathArgument makes sense to me

Bouke commented on Sources/Utility/ArgumentParser.swift:
```diff
@@ -383,6 +404,11 @@ public final class ArgumentParser {
     }
     
+    /// Add a parser with a subcommand name.
+    public func add(subparser command: String, parser: ArgumentParser) {
```
At the moment it is used to create the "root" argument parser, for generating any completions after `swift `: tools and compiler arguments.

aciidb0mb3r commented on Sources/Commands/SwiftPackageTool.swift:
```diff
@@ -455,8 +468,38 @@ public enum PackageMode: String, StringEnumArgument {
-extension InitMode: StringEnumArgument {}
-extension ShowDependenciesMode: StringEnumArgument {}
-extension DescribeMode: StringEnumArgument {}
+extension InitMode: StringEnumArgument {
```
As I said lets land all these changes separately and focus on ArgumentParser's ability to generate autocomplete scripts

Bouke replied:
Ok I've dropped the script generation commands from this PR. I've included information in the generated output on how to use the shell scripts.

Bouke commented:
Ok I've dropped the script generation commands from this PR. I've included information in the generated output on how to use the shell scripts. I've also rebased the PR on top of master.

Bouke commented:
What's the status on this one?

aciidb0mb3r commented on Sources/Commands/Options.swift:
```diff
@@ -42,4 +42,26 @@ public class ToolOptions {
+    }
+}
+
+struct PathArgument: ArgumentKind {
```
Can we get doc comments on what this is?

aciidb0mb3r replied:
We should add this in a separate PR and with `ArgumentParser` itself, rather than in Commands module.

Bouke replied:
As discussed on slack, I'll keep this in this PR for now.

aciidb0mb3r commented on Sources/Commands/Options.swift:
```diff
@@ -42,4 +42,26 @@ public class ToolOptions {
+}
+
+struct PathArgument: ArgumentKind {
+    let path: String
```
This should be `AbsolutePath` if its a path argument

Bouke replied:
We discussed that constructing the `AbsolutePath` could take place inside `PathArgument.init`.

aciidb0mb3r commented on Sources/Utility/ArgumentParser.swift:
```diff
@@ -53,6 +81,8 @@ extension String: ArgumentKind {
         self = try parser.associatedArgumentValue ?? parser.next()
     }
+
+    public static let completion = ShellCompletion.none
```
Lets make this consistent with others:
```swift
public static let completion: ShellCompletion = .none
```

aciidb0mb3r commented on Sources/Utility/ArgumentParser.swift:
```diff
@@ -193,7 +227,7 @@ public final class PositionalArgument<Kind>: ArgumentProtocol {
 ///
 /// Note: Only used for argument parsing purpose.
-fileprivate final class AnyArgument: ArgumentProtocol, CustomStringConvertible {
+internal final class AnyArgument: ArgumentProtocol, CustomStringConvertible {
```
We don't need to write internal

aciidb0mb3r commented on Sources/Utility/ArgumentParser.swift:
```diff
@@ -320,11 +354,11 @@ public final class ArgumentParser {
-    private var options = [AnyArgument]()
-    private var positionalArgs = [AnyArgument]()
+    internal var options = [AnyArgument]()
+    internal var positionalArgs = [AnyArgument]()
```
this should be `private(set) var ..` instead of internal var

Bouke commented:
I've rebased this PR on master

Bouke commented:
What's the status on this one?

aciidb0mb3r commented:
LGTM!
Please resolve conflicts and rebase then we can run swiftci and merge.
Thank you for your work on this!

Bouke commented:
Great! I've rebased the PR and tests successfully on my machine.

aciidb0mb3r commented:
@swift-ci please smoke test

aciidb0mb3r commented:
@swift-ci please smoke test
